src/Compilers/Core/Portable/Collections/TopologicalSort.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// A helper class that contains a topological sort algorithm.
+    /// </summary>
+    internal static class TopologicalSort
+    {
+        /// <summary>
+        /// Produce a topological sort of a given directed acyclic graph, given a set of nodes which include all nodes
+        /// that have no predecessors. Any nodes not in the given set, but reachable through successors, will be added
+        /// to the result. This is an iterative rather than recursive implementation, so it is unlikely to cause a stack
+        /// overflow.
+        /// </summary>
+        /// <typeparam name="TNode">The type of the node</typeparam>
+        /// <param name="nodes">Any subset of the nodes that includes all nodes with no predecessors</param>
+        /// <param name="successors">A function mapping a node to its set of successors</param>
+        /// <returns>A list of all reachable nodes, in which each node always precedes its successors</returns>
+        public static ImmutableArray<TNode> IterativeSort<TNode>(IEnumerable<TNode> nodes, Func<TNode, IEnumerable<TNode>> successors)
+        {
+            // First, count the predecessors of each node
+            PooledDictionary<TNode, int> predecessorCounts = PredecessorCounts(nodes, successors);
+
+            // Initialize the ready set with those nodes that have no predecessors
+            var ready = ArrayBuilder<TNode>.GetInstance();
+            foreach ((TNode node, int count) in predecessorCounts)
```

Comment:
Are we dependent on the order of keys in a hashtable? 

A completely stable sort  WRT nodes/successors is probably an overkill, but I wonder if we could be getting different results when sorting something like symbols and whether that is ok or we never do that.

Reply:
That is a good point. The sort should be deterministic. I will build a deterministically-ordered list of nodes and use that here.

Reply:
As pointed out in https://github.com/dotnet/roslyn/issues/23391#issuecomment-348169125, the order of `Dictionary<TKey, TValue>` is deterministic as long as items are only added to it and never removed. It is unclear if that guarantee applies to dictionaries that are cleared (via `Clear()`).

Reply:
hashes of the stuff we are going to put into the dictionary is not deterministic (bound nodes, symbols..) 
In many cases the hash would be, or derived from, object hashes. That can easily change between builds.

Reply:
@VSadov `Dictionary<TKey, TValue>` enumeration order equals the original insertion order until an item is removed and then another item is added, where the hashes have no impact on this rule.

Reply:
https://github.com/dotnet/coreclr/blob/e93926660c0e1f0aad452ced770e5859e49b1060/src/mscorlib/shared/System/Collections/Generic/Dictionary.cs#L1111

It looks like just going through buckets.

Reply:
We replace items in the dictionary here. I don't want to rely on Dictionary order in any case.

---
In reply to: [171730132](https://github.com/dotnet/roslyn/pull/25169#discussion_r171730132) [](ancestors = 171730132)

Reply:
What I mean is the Dictionary does not have any facility to remember the insertion order. It simply enumerates the internal bucket array where items are stored based on their truncated key hash values or whether there were collisions.  
In our case it is fairly nondeterministic.